src/mlpack/methods/reinforcement_learning/ppo_impl.hpp

```diff
@@ -0,0 +1,342 @@
+  PolicyType,
+  ReplayType
+>::PPO(TrainingConfig config,
+       ActorNetworkType actor,
+       CriticNetworkType critic,
+       PolicyType policy,
+       ReplayType replayMethod,
+       UpdaterType updater,
+       EnvironmentType environment):
+  config(std::move(config)),
+  actorNetwork(std::move(actor)),
+  criticNetwork(std::move(critic)),
+  actorUpdater(std::move(updater)),
+  #if ENS_VERSION_MAJOR >= 2
+  actorUpdatePolicy(NULL),
+  #endif
+  criticUpdater(std::move(updater)),
+  #if ENS_VERSION_MAJOR >= 2
+  criticUpdatePolicy(NULL),
+  #endif
+  policy(std::move(policy)),
+  replayMethod(std::move(replayMethod)),
+  environment(std::move(environment))
+{
+  // Set up actor and critic network.
+  if (actorNetwork.Parameters().is_empty())
+    actorNetwork.ResetParameters();
+
+  if (criticNetwork.Parameters().is_empty())
+    criticNetwork.ResetParameters();
+
+  #if ENS_VERSION_MAJOR == 1
+  this->criticUpdater.Initialize(criticNetwork.Parameters().n_rows,
+                                 criticNetwork.Parameters().n_cols);
+  #else
+  this->criticUpdatePolicy = new typename UpdaterType::template
+  Policy<arma::mat, arma::mat>(this->criticUpdater,
+                               criticNetwork.Parameters().n_rows,
+                               criticNetwork.Parameters().n_cols);
+  #endif
+
+  #if ENS_VERSION_MAJOR == 1
+  this->actorUpdater.Initialize(actorNetwork.Parameters().n_rows,
+                                actorNetwork.Parameters().n_cols);
+  #else
+  this->actorUpdatePolicy = new typename UpdaterType::template
+  Policy<arma::mat, arma::mat>(this->actorUpdater,
+                               actorNetwork.Parameters().n_rows,
+                               actorNetwork.Parameters().n_cols);
+  #endif
+
+  oldActorNetwork = actorNetwork;
+}
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType
+>
+PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::~PPO()
+{
+  #if ENS_VERSION_MAJOR >= 2
+  delete actorUpdatePolicy;
+  delete criticUpdatePolicy;
+  #endif
+}
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType
+>
+void PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::Update()
+{
+  arma::mat sampledStates;
+  std::vector<ActionType> sampledActions;
+  arma::colvec sampledRewards;
+  arma::mat sampledNextStates;
+  arma::icolvec isTerminal;
+
+  replayMethod.Sample(sampledStates, sampledActions, sampledRewards,
+                      sampledNextStates, isTerminal);
+
+  arma::rowvec discountedRewards(sampledRewards.n_rows);
+  arma::mat nextActionValues;
+  double values = 0.0;
+  criticNetwork.Predict(sampledNextStates, nextActionValues);
```

Comment:
@robotcator could you please explain the requirement of `criticNetwork.Predict` here, I think `nextActionValues` is not used anywhere in this part